### PR TITLE
Disallow side effects in asserts

### PR DESCRIPTION
**Pull request by charles-cooper**
### - What I did
Implemented #1150 - disallow side effects in asserts

### - How I did it
Added an `in_assertion` flag and check it wherever `is_constant` is checked

### - How to verify it
Run the tests

### - Description for the changelog
Disallow side effects in assert statements.

### CODE REVIEW

File: vyper/parser/parser.py [hunk 1]
```diff
@@ -342,7 +342,7 @@ def parse_func(code, sigs, origcode, global_ctx, _vars=None):
         global_ctx=global_ctx,
         sigs=sigs,
         return_type=sig.output_type,
-        is_constant=sig.const,
+        is_constant='a constant function' if sig.const else None,
```
Comment (jacqueswww) [inline vyper/parser/parser.py]: Rather use a boolean value, having a string like that is inconsistent with the other context flags.

File: vyper/parser/stmt.py [hunk 2]
```diff
@@ -322,7 +322,11 @@ def call(self):
             return external_call.make_external_call(self.stmt, self.context)
 
     def parse_assert(self):
+        tmp = self.context.is_constant # backup value
+        self.context.is_constant = 'an assert'
```
Comment (jacqueswww) [inline vyper/parser/stmt.py]: I believe it makes much more sense to have a separate flag/structure for 'in asssert expression' instead, see blockscope/in_assignment/in_for_loop -  using is_constant as a string is a bit hackish.

**Comment by charles-cooper** [comment]
@jacqueswww Thanks for the review! requested changes pushed

**Comment by jacqueswww** [comment]
Just flake8 fixes, and then we can merge.

**Comment by charles-cooper** [comment]
Ah yes, I always forget to run the linter. Lint removed!

**Pull request merged**